File: src/damposc.py
# -*- coding: utf-8 -*-
"""
Created on Tue May 30 17:24:20 2023
Conversion of damposc.m to python.
@author: cognotrend
"""

import numpy as np
import matplotlib.pyplot as plt
from scipy.linalg import expm
from collate import collate

np.random.seed(0)  # Set random seed for reproducibility

num_samples =50
delta_t = 1
m = 1 # kg
spring_coeff = 1 # N/m light spring
damp_coeff   = 5 # Ns/m medium damping

Phi = expm(np.array([[0, 1], [-spring_coeff/m, -damp_coeff/m]]) * delta_t)

# Model statistics:
q_cont = 0.01 # contuous process noise variance 
Q = q_cont*np.array([[delta_t**3/3.0, delta_t**2/2.0],[delta_t**2/2.0,delta_t]])
P0 = np.diag([0.01**2, 0.1**2])
R = 0.10**2
H = np.array([1, 0])
x_true = np.zeros((2, num_samples))

# True vs. assumed process noise:
# Make this parameter=1, if you want the truth and the model to have the same process
# noise stats.  Otherwise, make it any non-negative value you like for parametric studies.
process_noise_assumption_factor = 0.8  
w1_true = process_noise_assumption_factor * np.sqrt(Q[0, 0]) * np.random.randn(num_samples) 
w2_true = process_noise_assumption_factor * np.sqrt(Q[1, 1]) * np.random.randn(num_samples)
w_true = np.vstack((w1_true, w2_true))
x_true[:, 0] = np.array([1, 0]) + w_true[:, 0] # truth at timestep 0

for k in range(1, num_samples):
    x_true[:, k] = Phi.dot(x_true[:, k-1]) + process_noise_assumption_factor*w_true[:, k-1]

x_hat = np.zeros((2, num_samples))
prior_est = np.zeros((2, num_samples))
x_hat[:, 0] = np.array([1, 0])

P = np.zeros((2, 2, num_samples))
prior_P = np.zeros((2, 2, num_samples))
pre = np.zeros((2,num_samples))
post = np.zeros((2,num_samples))

# Pre-compute Measurements:
v = np.sqrt(R) * np.random.randn(num_samples)
z = x_true[0,:] + v

P[:, :, 0] = P0
post[0] = P[0, 0, 0]
pre[0] = P[0, 0, 0]


for k in range(1, num_samples):
    prior_est[:, k] = Phi.dot(x_hat[:, k-1])
    prior_P[:, :, k] = Phi.dot(P[:, :, k-1]).dot(Phi.T) + Q
    pre[0,k] = prior_P[0, 0, k]
    pre[1,k] = prior_P[1, 1, k]
    K = prior_P[:, :, k].dot(H.T) / (H.dot(prior_P[:, :, k]).dot(H.T) + R)
    x_hat[:, k] = prior_est[:, k] + K * (z[k] - H.dot(prior_est[:, k]))
    P[:, :, k] = (np.eye(2) - K.dot(H)).dot(prior_P[:, :, k]).dot((np.eye(2) - K.dot(H)).T) + K.dot(R).dot(K.T)
    post[0,k] = P[0, 0, k]
    post[1,k] = P[1, 1, k]

rms_pos = np.sqrt(np.mean((x_true[0, :] - x_hat[0, :])**2))
rms_vel = np.sqrt(np.mean((x_true[1, :] - x_hat[1, :])**2))
posQ = f'Pos Q={np.round(Q[0,0],6)}'
velQ = f'Vel Q={np.round(Q[1,1],6)}'
pos_init_est = f'Pos={np.round(x_hat[0,0],2)}'
vel_init_est = f'Vel={np.round(x_hat[1,0],2)}'
param_string = f'Model Init: {pos_init_est}, {vel_init_est}, {posQ}, {velQ},\n Pos P={np.round(P[0,0,0],4)}, Vel P={np.round(P[1,1,0],4)}, R={np.round(R,6)}'

fig, axs = plt.subplots(2, 1, figsize=(8, 8))
times = delta_t*np.arange(1, num_samples+1)
axs[0].plot(times, x_true[0, :], 'b--*', label='Truth')
axs[0].plot(times, z, 'r--o', label='Measurements')
axs[0].plot(times, x_hat[0, :], 'd:k', label='Estimate')
axs[0].legend(loc='lower left')
axs[0].set_xlabel(f'Time (s) $\Delta_t$ ={delta_t}')
axs[0].set_ylabel('Position (m)')
axs[0].set_title('Kalman Estimator: Damped Harmonic Oscillator\n'+param_string)
str_text = f"RMS Pos Err: {np.round(rms_pos,4)}"
axs[0].text(0.5, 0.1, str_text, transform=axs[0].transAxes)
axs[1].plot(times, x_true[1, :], 'b--*', label='Truth')
axs[1].plot(times, x_hat[1, :], 'd:k', label='Estimate')
axs[1].legend(loc='lower left')
axs[1].set_xlabel(f'Time (s), $\Delta t$={delta_t}') 
axs[1].set_ylabel('Velocity (m/sec)')
str_text = f"RMS Vel Err: {np.round(rms_vel,4)}"
axs[1].text(0.5, 0.1, str_text, transform=axs[1].transAxes)

print('Error:')
print(x_hat[:, -1] - x_true[:, -1])

collated = collate(pre[0,:], post[0,:])

new_time = collate(times, times)
plt.figure(2)
plt.plot(new_time, collated)
plt.title('Variance in Position Estimate (w/ Pos Sensor Only)\n'+param_string)
plt.xlabel(f'Time(s), $\Delta t$={delta_t} ')
plt.ylabel('Variance (m^2)')

plt.show()

collated = collate(pre[1,:], post[1,:])

new_time = collate(times, times)
plt.figure(3)
plt.plot(new_time, collated)
plt.title('Variance in Velocity Estimate (w/ Pos Sensor Only)\n'+param_string)
plt.xlabel(f'Time(s), $\Delta t$={delta_t} ')
plt.ylabel('Variance (m^2)')

plt.show()

File: src/matlab/damposc.m
% damposc.m  Damped harmonic oscillator
% See Harmonic Oscillator examtple in Day 1, Segment 2
% 
clear
num_samples = 30;

delta_t = 1;
m = 1;    % kg
c = 0.1;  % netwon/meter  Spring constant
b = 0.1; % newton sec / meter  dampling coefficient

Phi = expm([0 1; -b/m -c/m]*delta_t);

Q = [0.00001 0; 0 0.000001];
%Q = zeros(2);
P0 = [0.01 0; 0 0.01];
R=0.1^2;
H=[1 0];
w1 = sqrt(Q(1,1))*randn(1,num_samples);
w2 = sqrt(Q(2,2))*randn(1,num_samples);

w = [w1;w2];
x_true = zeros(2,num_samples);
x_true(:,1) = [1;0] + w(:,1);

for k=2:num_samples
    x_true(:,k) = Phi*x_true(:,k-1) + w(:,k-1);
end


x_hat = zeros(2,num_samples);
prior_est = zeros(2,num_samples);
x_hat(:,1)= [1;0];

P = zeros(2,2,num_samples);
prior_P = zeros(2,2,num_samples);

pre = zeros(1,num_samples);
post = zeros(1,num_samples);
z = zeros(1,num_samples);

P(:,:,1)=P0;
post(1) = P(1,1,1);
pre(1) = P(1,1,1);
z(1) = x_true(1,1)+sqrt(R)*randn();

for k=2:num_samples
    prior_est(:,k) = Phi*x_hat(:,k-1); % No dynamics and noise noise because this is an estimate!
    prior_P(:,:,k) = Phi*P(:,:,k-1)*Phi' + Q;
    pre(k) = prior_P(1,1,k);
    K = prior_P(:,:,k)*H'*inv(H*prior_P(:,:,k)*H' + R);
    z(k) = x_true(1,k)+sqrt(R)*randn();
    x_hat(:,k) = prior_est(:,k) + K*(z(k)-H*prior_est(:,k));

    P(:,:,k) = (eye(2)-K*H)*prior_P(:,:,k)*(eye(2)-K*H)' + K*R*K';
    post(k) = P(1,1,k);
end

% RMS Errors:

rms_pos = sqrt(mean((x_true(1,:)-x_hat(1,:)).^2));
rms_vel = sqrt(mean((x_true(2,:)-x_hat(2,:)).^2));

figure(1)
subplot(2,1,1), plot(1:num_samples,x_true(1,:),'b--*',1:num_samples,z,'r--o',1:num_samples,x_hat(1,:),'d:k');
legend('Truth','Measurements','Estimate','Location','southeast');
xlabel('Time')
ylabel('Position (m)')
title('Kalman Estimator: Damped Harmonic Oscillator')
str = {['Meas Noise Variance: ',num2str(R)],['RMS Pos Err: ',num2str(rms_pos)]};
annotation('textbox', [0.7,0.8,0.15,0.05],...
           'String', str);
subplot(2,1,2), plot(1:num_samples,x_true(2,:),'b--*',1:num_samples,x_hat(2,:),'d:k');       
legend('Truth','Estimate','Location','southeast');
xlabel('Time (s)')
ylabel('Velocity (m/sec)')
str = {['RMS Vel Err: ',num2str(rms_vel)]};
annotation('textbox', [0.7,0.4,0.15,0.03],...
           'String', str);

display('Error: ')
display(x_hat(:,num_samples)-x_true(:,num_samples))

% To get a sawtooth plot, you can use the attached function I wrote called collate.m.
% If you have two vectors of variances, v1, v2, and one is supposed to be before measurement update and one after, you create
% a new vector twice in length by v=collate(v1,v2).   Then, if t is a the time vector for the period in question, you collate it with itself by
%new_time = collate(t,t).   Then you can plot v against new_time:   plot(new_time, v).

collated = collate(pre,post);
t=1:num_samples;
new_time = collate(t,t);
figure(2)
plot(new_time,collated);
title('Variance in Position (with Process Noise)')
xlabel('Time')
ylabel('Variance (m^2)');




